### PR TITLE
Share schema definition using JsonSchemaValidator instance

### PR DESCRIPTION
**Pull request by yufeih**
Currently `JsonSchemaValidation` and `JsonSchemaTransform` creates a new _definition_ table for each file, by creating a `Validator` | `Transformer` object, this _definition_ table is creates once per schema.

### CODE REVIEW

File: src/docfx/build/page/BuildPage.cs [hunk 1]
```diff
@@ -183,18 +183,18 @@ private static (List<Error> errors, Schema schema, OutputModel model)
                 throw Errors.SchemaNotFound(file.Mime).ToException();
             }
 
-            var jsonSchema = TemplateEngine.GetJsonSchema(file.Schema);
-            if (jsonSchema is null)
+            var (jsonSchema, schemaValidator, schemaTransformer) = TemplateEngine.GetJsonSchema(file.Schema);
```
Comment (OsmondJiang) [inline src/docfx/build/page/BuildPage.cs]: remove jsonSchema, it's no used anymore?

File: src/docfx/build/xref/XrefMap.cs [hunk 2]
```diff
@@ -475,13 +475,13 @@ private static (List<Error> errors, IReadOnlyList<InternalXrefSpec> specs) LoadS
             }
 
             var errors = new List<Error>();
-            var jsonSchema = TemplateEngine.GetJsonSchema(file.Schema);
-            if (jsonSchema is null)
+            var (schema, schemaValidator, schemaTransformer) = TemplateEngine.GetJsonSchema(file.Schema);
```
Comment (OsmondJiang) [inline src/docfx/build/xref/XrefMap.cs]: same here

File: src/docfx/lib/schema/JsonSchemaDefinition.cs [hunk 3]
```diff
@@ -22,24 +22,24 @@ public JsonSchemaContext(JsonSchema root)
             _definitions.Add("#", root);
         }
 
-        public JsonSchema GetDefinition(JsonSchema jsonSchema)
-            => GetDefinitionCore(jsonSchema, new HashSet<string>());
+        public JsonSchema GetDefinition(JsonSchema subSchema)
```
Comment (OsmondJiang) [inline src/docfx/lib/schema/JsonSchemaDefinition.cs]: it's the root json schema, not sub schema, why we change the name?

File: src/docfx/lib/schema/JsonSchemaDefinition.cs [hunk 4]
```diff
@@ -22,24 +22,24 @@ public JsonSchemaContext(JsonSchema root)
...
         {
-            if (jsonSchema != null &&
-                !string.IsNullOrEmpty(jsonSchema.Ref) &&
-                recursions.Add(jsonSchema.Ref))
+            if (subSchema != null &&
+                !string.IsNullOrEmpty(subSchema.Ref) &&
+                recursions.Add(subSchema.Ref))
             {
-                if (_definitions.TryGetValue(jsonSchema.Ref, out var schema))
+                if (_definitions.TryGetValue(subSchema.Ref, out var schema))
```
Comment (OsmondJiang) [inline src/docfx/lib/schema/JsonSchemaDefinition.cs]: that's the subSchema

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 5]
```diff
@@ -8,55 +8,72 @@
...
+            void TransformXrefCore(JsonSchema schema, JValue value)
+            {
+                extensions[value.Path] = new Lazy<JValue>(
+                    () => TransformValue(schema, file, context, value, errors, buildChild: null),
+                    LazyThreadSafetyMode.PublicationOnly);
+            }
         }
 
-        private static void Traverse(JsonSchemaContext jsonSchemaContext, JsonSchema schema, JToken token, Action<JsonSchema, JValue> transformScalar)
+        private void Traverse(JsonSchema subSchema, JToken token, Action<JsonSchema, JValue> action)
```
Comment (OsmondJiang) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: same here

Comment (OsmondJiang) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: we may have TransformArray action, so suggest to keep the origin action name

File: src/docfx/template/TemplateEngine.cs [hunk 6]
```diff
@@ -17,7 +17,8 @@ namespace Microsoft.Docs.Build
     internal class TemplateEngine
     {
         private static readonly string[] s_resourceFolders = new[] { "global", "css", "fonts" };
-        private static readonly ConcurrentDictionary<string, JsonSchema> _jsonSchemas = new ConcurrentDictionary<string, JsonSchema>();
+        private static readonly ConcurrentDictionary<string, Lazy<(JsonSchema, JsonSchemaValidator, JsonSchemaTransformer)>> _jsonSchemas
+                          = new ConcurrentDictionary<string, Lazy<(JsonSchema, JsonSchemaValidator, JsonSchemaTransformer)>>();
```
Comment (OsmondJiang) [inline src/docfx/template/TemplateEngine.cs]: don't need expose it

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 7]
```diff
@@ -8,55 +8,72 @@
...
     {
-        public static (List<Error> errors, JToken token) TransformContent(Document file, Context context, JsonSchema schema, JToken token, Action<Document> buildChild)
+        private readonly JsonSchema _schema;
+        private readonly JsonSchemaDefinition _definitions;
+
+        public JsonSchemaTransformer(JsonSchema schema)
         {
-            var errors = new List<Error>();
+            _schema = schema;
+            _definitions = new JsonSchemaDefinition(schema);
```
Comment (live1206) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: We are creating a new instance of `JsonSchemaDefinition` each for `JsonSchemaTransformer` and `JsonSchemaValidator`, Is it possible to share 1 instance for both?

Comment (yufeih) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: just to make the interface simple, one extra instance is likely not going to cause perf problems, an instance per file is very likely.

**Pull request merged**